### DISCUSSION TITLE
Improve testing infrastructure

### DISCUSSION THREAD

**Discussion by KelvinChung2000** [FPGA-Research/FABulous#183] [Ideas]
Start using a testing framework, for example, [pytest](https://docs.pytest.org/en/8.2.x/), to make testing more scaleable and maintainable.

**Comment by IAmMarcelJung** [comment]
The recent issues showed that we definitely need something like this (was about to open a discussion until I saw that this already exists). We should test both the python code and the HDL code more extensively.

**Reply by KelvinChung2000** [reply]
I have started working on it. https://github.com/KelvinChung2000/FABulous/tree/test-ic Feel free to contribute.

**Reply by IAmMarcelJung** [reply]
Cool! Will try to do that once I have more time, which will probably take until April or May unfortunately...
